### DISCUSSION TITLE
Why does SVD no longer determine whether the matrix contains infinite numbers?

### DISCUSSION THREAD

**Discussion by mayuyuace** [jax-ml/jax#23234] [General] [answered]
https://github.com/google/jax/commit/b2654c08f989c1bbb758758214034625ef3a0b76?diff=split&w=0
Is this a hardware improvement or was it determined in XLA?
If it was a change in XLA, could you tell me the corresponding commits?

**Comment by jakevdp** [comment]
Hi - thanks for the question! It looks like this change was purely on the JAX side, and not associated with any XLA commits. Pinging for more info @rmlarsen, who authored the change.

**Comment by mayuyuace** [comment]
I found that NAN is checked at the end of the svd calculation, but before that, it might have returned early here:
https://github.com/google/jax/blob/be6e1549c344a90a57deb777cf935b3a432805f9/jax/_src/lax/svd.py#L214 
However, the function _svd_tall_and_square_input called here did not check NAN on the input.
I think the same check on input is needed in this function

**Comment by rmlarsen** [comment, marked as answer]
When using JIT compilation there is no such thing as "returning early", so we decided to get of rid of the special casing to simplify the code and slightly speed up the common case.

**Reply by mayuyuace** [reply]
Thank you, I understand that although not checking for Nan values ​​increases the overhead when the input contains Nan values, in general this will reduce a certain amount of overhead.

**Comment by Zantares** [comment]
Hi, is there any exact performance or case available for this optimization? I tried jax v0.4.26 (before opt) and jax v0.4.28 (after opt) with public jax UT `svd_test.py` on A100, but I got strange performance as below:

| Case name | 0.4.26 duration | 0.4.28 duration |
| ------------ | ------------------ | ------------------ |
| SvdTest.testSvdvals0 | 2.309 s | 2.267 s |
| SvdTest.testSvdWithRectangularInput0 | 29.072s | 39.460 s |

I looped each case **100 times** to watch the performance, but there's no performance improvement after opt...

**Reply by rmlarsen** [reply]
As mentioned above, the expected improvement is tiny, so perhaps consistent with your 2.309 -> 2.267s change? The change was mostly in the interest of simplifying the code. However, it looks like you are seeing a large performance regression on rectangular inputs. What is that test case exactly?

**Reply by Zantares** [reply]
It's a sub-case in public JAX UT `svd_test.py`: https://github.com/google/jax/blob/main/tests/svd_test.py#L68
I modified the function body by adding a 100 times loop, to check the performance:
```python
  def testSvdWithRectangularInput(self, m, n, log_cond, full_matrices):
    """Tests SVD with rectangular input."""
    for i in range(100):
      with jax.default_matmul_precision('float32'):
        a = np.random.uniform(
            low=0.3, high=0.9, size=(m, n)).astype(_SVD_TEST_DTYPE)
        ...
```

Run cmd: `python tests/svd_test.py SvdTest.testSvdWithRectangularInput0`

**Reply by rmlarsen** [reply]
Thank you. I'll look into it.

**Reply by rmlarsen** [reply]
@Zantares what happens if you install JAX v0.4.26 and copy the files in jax/_src/lax/linalg from  v0.4.28 there? This should verify that the changes to the SVD code are to blame and not some other unrelated change in the GPU compiler stack.

**Reply by rmlarsen** [reply]
Actually, I don't think the implementation in  jax/_src/lax/linalg is used for SVD on GPU. @jakevdp can you verify?

**Reply by jakevdp** [reply]
The GPU svd lowering is found here: https://github.com/google/jax/blob/ea5fd29b90b244e29e382357bf641bbd542f7fab/jax/_src/lax/linalg.py#L2189-L2192

The changes referenced in this question (in `jax/_src/lax/svd.py` and `jax/_src/lax/qdwh.py`) are only used for the TPU lowering rule I believe.

**Reply by rmlarsen** [reply]
(Just for completeness: `jax/_src/lax/eigh.py`  is also called by `jax/_src/lax/svd.py`.)

**Reply by rmlarsen** [reply]
But yeah, changes to the GPU SVD seem unrelated.

**Reply by Zantares** [reply]
> @Zantares what happens if you install JAX v0.4.26 and copy the files in jax/_src/lax/linalg from v0.4.28 there? This should verify that the changes to the SVD code are to blame and not some other unrelated change in the GPU compiler stack.

Thanks for the suggestion, I will try it.

**Comment by Zantares** [comment]
Hi @rmlarsen , @jakevdp , I did below changes and retested again:
* Revert the svd opt in jax v0.4.28 
* Cherry-pick svd opt to jax v0.4.26 and handle the conflict manually

It seems that the svd opt still impacted part of the GPU performance:

| Case name | 0.4.26 duration | 0.4.28 duration | 0.4.28 W/O opt duration | 0.4.26 W/ opt duration |
| ------------ | ----------------- | ---------------- | ------------------------------- | ------------------------------ |
| SvdTest.testSvdWithRectangularInput0 | 29.072 s | 39.460 s | 34.407 s | 35.394 s |

As you mentioned, jax v0.4.28 has other changes which impacted performance, but this svd opt still brought additional ~5 s regression here.

**Reply by rmlarsen** [reply]
what exactly did you revert? The changes in eigh.py, svd.py, and qdwh.py should have no effect on GPU.

**Reply by Zantares** [reply]
I reverted/cherry-picked this commit b2654c0 https://github.com/google/jax/commit/b2654c08f989c1bbb758758214034625ef3a0b76?diff=split&w=0

**Reply by Zantares** [reply]
I feel like that the case runs on GPU because:
* The call stack will be printed in `_qdwh()` when test it on A100:
```log
  File "/mnt/tmp/home/sparse/tenglu/jax/tests/svd_test.py", line 84, in testSvdWithRectangularInput
    actual_u, actual_s, actual_v = svd.svd(a, full_matrices=full_matrices)
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/traceback_util.py", line 179, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/pjit.py", line 298, in cache_miss
    outs, out_flat, out_tree, args_flat, jaxpr, attrs_tracked = _python_pjit_helper(
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/pjit.py", line 167, in _python_pjit_helper
    _infer_params(jit_info, args, kwargs)
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/pjit.py", line 581, in _infer_params
    jaxpr, consts, out_shardings, out_layouts_flat, attrs_tracked = _pjit_jaxpr(
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/pjit.py", line 1202, in _pjit_jaxpr
    jaxpr, final_consts, out_type, attrs_tracked = _create_pjit_jaxpr(
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/linear_util.py", line 350, in memoized_fun
    ans = call(fun, *args)
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/pjit.py", line 1146, in _create_pjit_jaxpr
    jaxpr, global_out_avals, consts, attrs_tracked = pe.trace_to_jaxpr_dynamic(
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/profiler.py", line 335, in wrapper
    return func(*args, **kwargs)
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/interpreters/partial_eval.py", line 2322, in trace_to_jaxpr_dynamic
    jaxpr, out_avals, consts, attrs_tracked = trace_to_subjaxpr_dynamic(
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/interpreters/partial_eval.py", line 2344, in trace_to_subjaxpr_dynamic
    ans = fun.call_wrapped(*in_tracers_)
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/linear_util.py", line 192, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/lax/svd.py", line 219, in svd
    u_out, s_out, v_out = _svd_tall_and_square_input(
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/traceback_util.py", line 179, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/pjit.py", line 298, in cache_miss
    outs, out_flat, out_tree, args_flat, jaxpr, attrs_tracked = _python_pjit_helper(
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/pjit.py", line 167, in _python_pjit_helper
    _infer_params(jit_info, args, kwargs)
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/pjit.py", line 581, in _infer_params
    jaxpr, consts, out_shardings, out_layouts_flat, attrs_tracked = _pjit_jaxpr(
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/pjit.py", line 1202, in _pjit_jaxpr
    jaxpr, final_consts, out_type, attrs_tracked = _create_pjit_jaxpr(
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/linear_util.py", line 350, in memoized_fun
    ans = call(fun, *args)
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/pjit.py", line 1146, in _create_pjit_jaxpr
    jaxpr, global_out_avals, consts, attrs_tracked = pe.trace_to_jaxpr_dynamic(
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/profiler.py", line 335, in wrapper
    return func(*args, **kwargs)
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/interpreters/partial_eval.py", line 2322, in trace_to_jaxpr_dynamic
    jaxpr, out_avals, consts, attrs_tracked = trace_to_subjaxpr_dynamic(
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/interpreters/partial_eval.py", line 2344, in trace_to_subjaxpr_dynamic
    ans = fun.call_wrapped(*in_tracers_)
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/linear_util.py", line 192, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/lax/svd.py", line 72, in _svd_tall_and_square_input
    u_p, h, _, _ = lax.linalg.qdwh(
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/traceback_util.py", line 179, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/pjit.py", line 298, in cache_miss
    outs, out_flat, out_tree, args_flat, jaxpr, attrs_tracked = _python_pjit_helper(
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/pjit.py", line 167, in _python_pjit_helper
    _infer_params(jit_info, args, kwargs)
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/pjit.py", line 581, in _infer_params
    jaxpr, consts, out_shardings, out_layouts_flat, attrs_tracked = _pjit_jaxpr(
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/pjit.py", line 1202, in _pjit_jaxpr
    jaxpr, final_consts, out_type, attrs_tracked = _create_pjit_jaxpr(
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/linear_util.py", line 350, in memoized_fun
    ans = call(fun, *args)
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/pjit.py", line 1146, in _create_pjit_jaxpr
    jaxpr, global_out_avals, consts, attrs_tracked = pe.trace_to_jaxpr_dynamic(
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/profiler.py", line 335, in wrapper
    return func(*args, **kwargs)
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/interpreters/partial_eval.py", line 2322, in trace_to_jaxpr_dynamic
    jaxpr, out_avals, consts, attrs_tracked = trace_to_subjaxpr_dynamic(
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/interpreters/partial_eval.py", line 2344, in trace_to_subjaxpr_dynamic
    ans = fun.call_wrapped(*in_tracers_)
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/linear_util.py", line 192, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/lax/qdwh.py", line 243, in qdwh
    u, h, num_iters, is_converged = _qdwh(x, m, n, is_hermitian, max_iterations,
  File "/mnt/tmp/home/sparse/tenglu/jax/jax/_src/lax/qdwh.py", line 128, in _qdwh
    traceback.print_stack()
```
* GPU utilization is changed from 0% to 6% after running this case

I'm not a Nvidia GPU expert. In fact we found this issue in our own PJRT device and used A100 to verify it again. Please tell us if anything is wrong and we would like to recheck it, thanks!

**Reply by rmlarsen** [reply]
That's surprising @jakevdp  I'm afraid don't know too much about JAX on GPU. I think the JAX team might have to help here.

**Reply by rmlarsen** [reply]
@Zantares  @jakevdp FWIW: I notice that NVIDIA also has an SVD based on polar decomposition, which is much faster than the _gesvd. We might consider providing an option to lower to that in JAX.

https://docs.nvidia.com/cuda/cusolver/index.html#cusolverdnxgesvdp

**Reply by jakevdp** [reply]
That test is calling `jax._src.lax.svd.svd()` directly, so if you're running the test suite on GPU, that code will run on GPU.

**Reply by Zantares** [reply]
> That test is calling `jax._src.lax.svd.svd()` directly, so if you're running the test suite on GPU, that code will run on GPU.

Do you mean that we should not pay much attention on this case since it's a special scenario?

**Reply by jakevdp** [reply]
What I mean is that the reason this code is being run on GPU is that it is being called directly. If that's similar to your actual usage, then it could still be problematic.

**Reply by rmlarsen** [reply]
@jakevdp thanks for figuring that out. This code was not optimized for GPU. It probably could be, but that was not the goal.

**Reply by rmlarsen** [reply]
I would recommend calling the path that gets lowered to the cuSolver SVD, which has been carefully optimized by NVIDIA.

**Reply by Zantares** [reply]
Got it, now I understand this question completely, thanks!

**Reply by Zantares** [reply]
One last question, is possible to add TPU device limitation here to fix this issue first?

**Reply by jakevdp** [reply]
There is no way to hit this code path on anything but TPU via public JAX APIs. The only reference to it is here, in the TPU-specific lowering rule: https://github.com/google/jax/blob/7d2f0a75c1c10aeae63ed1a0ced17b7aeb979989/jax/_src/lax/linalg.py#L2193-L2197

The test file runs it on GPU because it calls the private API directly. I suppose we could mark the test to be skipped unless running on a TPU backend, but that would be a net loss in test coverage, so I'm not sure it's warranted.
